state/etcd/etcd.go

```diff
@@ -0,0 +1,499 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/components-contrib/state"
+	"github.com/dapr/components-contrib/state/utils"
+	"github.com/dapr/kit/logger"
+	"github.com/dapr/kit/ptr"
+)
+
+const (
+	tlsEnable = "enable"
+)
+
+// Etcd is a state store implementation for Etcd.
+type Etcd struct {
+	state.DefaultBulkStore
+	client        *clientv3.Client
+	keyPrefixPath string
+	features      []state.Feature
+	logger        logger.Logger
+}
+
+type etcdConfig struct {
+	Endpoints     string `json:"endpoints"`
+	KeyPrefixPath string `json:"keyPrefixPath"`
+	// TLS
+	TlsEnable string `json:"tlsEnable"`
+	Ca        string `json:"ca"`
+	Cert      string `json:"cert"`
+	Key       string `json:"key"`
+}
+
+// NewEtcdStateStore returns a new etcd state store.
+func NewEtcdStateStore(logger logger.Logger) state.Store {
+	s := &Etcd{
+		logger:   logger,
+		features: []state.Feature{state.FeatureETag, state.FeatureTransactional},
+	}
+	s.DefaultBulkStore = state.NewDefaultBulkStore(s)
+
+	return s
+}
+
+// Init does metadata and config parsing and initializes the
+// Etcd client.
+func (e *Etcd) Init(metadata state.Metadata) error {
+	etcdConfig, err := metadataToConfig(metadata.Properties)
+	if err != nil {
+		return fmt.Errorf("couldn't convert metadata properties: %w", err)
+	}
+
+	e.client, err = e.ParseClientFromConfig(etcdConfig)
+	if err != nil {
+		return fmt.Errorf("initializing etcd client: %w", err)
+	}
+
+	e.keyPrefixPath = etcdConfig.KeyPrefixPath
+
+	return nil
+}
+
+func (e *Etcd) ParseClientFromConfig(etcdConfig *etcdConfig) (*clientv3.Client, error) {
+	endpoints := strings.Split(etcdConfig.Endpoints, ",")
+	if len(endpoints) == 0 || endpoints[0] == "" {
+		return nil, fmt.Errorf("endpoints required")
+	}
+
+	var tlsConfig *tls.Config
+	if etcdConfig.TlsEnable == tlsEnable {
```

Comment:
Rather than using a constant, please use a boolean.


```suggestion
	if utils.IsTruthy(etcdConfig.TlsEnable) {
```